pkg/endpoint/bpf.go

```diff
@@ -1183,6 +1205,25 @@ func (e *Endpoint) applyPolicyMapChanges() error {
 		}
 	}
 
+	if e.desiredPolicy.GetPolicyMap().Len() > policymap.MaxEntries {
```

Comment:
Even if the total number is less than MaxEntries, the insertion below may still fail, because insertions are performed before removals. But I guess that is fine, because then removals are performed (as we don't return on first error), which means that eventually insertions should succeed.

Reply:
There is not insertion below. New maps are created and replace the old maps.

Reply:
Ah, I was referring to the case in which `e.desiredPolicy.GetPolicyMap().Len()` is less than (but close to) MaxEntries, so the endpoint doesn't enter into lock-down mode. But a few insertions may still fail during the first attempt if they cause an overflow, compensated by the subsequent removal of other entries (if I'm not missing anything).

Reply:
Good catch. I added a section after lockdown mode that deletes changes first if the adds will overflow the map (temporarily).

Reply:
If I'm not mistaken, there could still be one issue caused by the explicit denies, when exiting from lockdown mode. Let's say `desiredPolicy.GetPolicyMap().Len() = MAX-3`, which theoretically fits into the map. But given that there are the additional 6 deny entries, the map will overflow, causing `applyPolicyMapChangesLocked` to return an error. And then the extra deny entries will never be removed, as the reconciliation is [performed afterwards](https://github.com/cilium/cilium/blob/pr%2Fnathanjsweet%2Fadd-lockdown-mode-for-policy-map-overflows/pkg/endpoint/bpf.go#L1353) only if the previous operation succeeded.

Reply:
> when exiting from lockdown mode. Let's say desiredPolicy.GetPolicyMap().Len() = MAX-3, which theoretically fits into the map. But given that there are the additional 6 deny entries, the map will overflow, causing applyPolicyMapChangesLocked to return an error.

The logic, as written, will delete entries first if the adds put `realizedPolicy` (which will account for the 3 entries) over the limit.

Reply:
Doesn't that account for desired policy changes only, while the reconciliation between desired and realized is performed in a subsequent stage? Otherwise I'm missing the purpose of [`syncPolicyMapsWith`](https://github.com/cilium/cilium/blob/pr%2Fnathanjsweet%2Fadd-lockdown-mode-for-policy-map-overflows/pkg/endpoint/bpf.go#L1353).